sourced/ml/algorithms/uast_ids_to_bag.py

```diff
@@ -1,46 +1,81 @@
 from collections import defaultdict
 
-from sourced.ml.algorithms import TokenParser
+import bblfsh
+
+from sourced.ml.algorithms import TokenParser, NoopTokenParser
 
 
 class FakeVocabulary:
     def __getitem__(self, item):
         return item
 
 
-class UastIds2Bag:
+class Uast2BagBase:
     """
-    Converts a UAST to a bag-of-identifiers.
+    Base class to convert UAST to bag of anything.
     """
-    def __init__(self, vocabulary, token_parser=None):
+    def __call__(self, uast):
         """
-        :param vocabulary: The mapping from tokens to bag keys. If None, no mapping is performed.
-        :param token_parser: Specify token parser if you want to use a custome one. \
-            :class:'TokenParser' is used if it is not specified.
+        Override this function with converting algorithm.
+
+        :param uast: The UAST root node.
+        :return: bag
         """
-        self._vocabulary = FakeVocabulary() if vocabulary is None else vocabulary
-        self._token_parser = TokenParser() if token_parser is None else token_parser
+        raise NotImplemented
+
+
+class UastTokens2Bag(Uast2BagBase):
+    """
+    Converts a UAST to a weighed bag of tokens via xpath.
+    """
+
+    XPATH = None  # Should be overridden in child class
+
+    def __init__(self, token2index=None, token_parser=None):
+        """
+        :param xpath: The libuast xpath query to filter nodes with needed tokens.
+        :param token2index: The mapping from tokens to bag keys. If None, no mapping is performed.
+        :param token_parser: Specify token parser if you want to use a custom one. \
+            :class:'NoopTokenParser' is used if it is not specified.
+        """
+        self._token2index = FakeVocabulary() if token2index is None else token2index
+        self._token_parser = NoopTokenParser() if token_parser is None else token_parser
 
     @property
-    def vocabulary(self):
-        return self._vocabulary
+    def token2index(self):
```

Comment:
hard to understand what is `vocabulary` means. Easier to get it if a name is `token2index`.